addons/stock/models/stock_move.py

```diff
@@ -1014,10 +1014,12 @@ def _keys_out_sorted(ml):
     def _action_cancel(self):
         if any(move.state == 'done' for move in self):
             raise UserError(_('You cannot cancel a stock move that has been set to \'Done\'.'))
+        self = self.filtered(lambda x: x.state != 'cancel')
+        # self cannot contain moves that are either cancelled or done, therefore we can safely
+        # unlink all associated move_line_ids
+        self.mapped('move_line_ids').unlink()
```

Comment:
If you can't have canceled or done moves here, the only extra processing you do in `_do_unreserve` is[*]:

https://github.com/odoo/odoo/blob/587e90fe18900037055fde2086597f6957e30a22/addons/stock/models/stock_move.py#L500

Is it costly? If not, I think it's better to stick to `_do_unreserve`, to avoid breaking potential overrides.

[*] I believe `state` is already in cache since you filtered.

Reply:
It's not costly, it's just a useless call

Reply:
In this case, I would remove the duplicated code from _action_cancel and return the move in _do_unreserve. It makes sense to unreserve before cancel a move and possible override could exist in custom code as nim said.